ports/nrf: Implement NVMC HAL.

aykevl opened the pull request:
This is only a library for flash access. Actual file system support will be added later.

See #96

aykevl committed:
ports/nrf: Implement NVMC HAL.

This is only a library for flash access. Actual file system support will
be added later.

glennrub commented on ports/nrf/hal/hal_nvmc.c:
```diff
@@ -0,0 +1,204 @@
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+
+    // Write all integers to flash.
+    for (int i=0; i<len; i++) {
```
some spaces around operators?

glennrub commented on ports/nrf/hal/hal_nvmc.c:
```diff
@@ -0,0 +1,204 @@
+    uint32_t value = 0xffffff00 | b;
+
+    // Rotate bits in value to an aligned position.
+    for (int i = dest - dest_aligned; i; i--) {
```
is this extra i; necessary?

aykevl replied:
You mean the middle argument to the for loop (the condition)? Yes, otherwise it won't stop looping. I'll change it to `i > 0` or something similar for clarity.

glennrub commented on ports/nrf/hal/hal_nvmc.h:
```diff
@@ -0,0 +1,68 @@
+#error Unknown chip
+#endif
+
+#define HAL_NVMC_IS_PAGE_ALIGNED(addr) ((uint32_t)(addr) & (HAL_NVMC_PAGESIZE-1))
```
space around operator -

glennrub commented on ports/nrf/hal/hal_nvmc.c:
```diff
@@ -0,0 +1,204 @@
+        hal_nvmc_write_words((uint32_t*)dest, (const uint32_t*)buf, len/4);
+        dest += len & ~0b11;
+        buf  += len & ~0b11;
+        len &= 0b11;
```
aligned with the others to keep consitency?

glennrub commented on ports/nrf/hal/hal_nvmc.c:
```diff
@@ -0,0 +1,204 @@
+    // Now the start of the buffer is aligned. Write as many words as
+    // possible, as that's much faster than writing bytes.
+    if (len / 4 && ((uint32_t)buf & 0b11) == 0) {
+        hal_nvmc_write_words((uint32_t*)dest, (const uint32_t*)buf, len/4);
```
spacing around /

glennrub commented on ports/nrf/hal/hal_nvmc.c:
```diff
@@ -0,0 +1,204 @@
+    uint32_t value = 0xffffff00 | b;
+
+    // Rotate bits in value to an aligned position.
+    for (int i = dest - dest_aligned; i; i--) {
```
is this extra i; necessary?

glennrub commented:
Thanks for the contribution, Ayke! I have looked a bit at your work before on your working branch, and this looks very good from my point of view. Just a few style comments from me.

aykevl commented:
Done. I applied the formatting changes, changed some code removing a for loop (reducing code size a little) and changed the global variable to a method (affecting the generated code but producing the same code size). Afterwards I tested whether it still works, of course ;)

glennrub commented:
Thanks! I like the update you did regarding the old global variable, exposing it through function call to the hal_nvmc instead.
